RA-1121/OMRSJS-5 migrate to webpack

PawelGutkowski opened the pull request.

PawelGutkowski commented:
okk, I've forgot about tests

PawelGutkowski commented:
I'm not able to make bitHound - Code test pass, as it fails on Yeoman placeholders.

psbrandt commented on README.md:
```diff
@@ -14,9 +14,9 @@
-  - [x] Local deploy with Gulp
-  - [x] Package management with [Bower](http://bower.io/)
+  - [x] Production build with [Webpack](https://webpack.github.io/)
+  - [x] Local deploy with [Webpack](https://webpack.github.io/)
```
This is minor, but we probably only need the first Webpack link :).

psbrandt commented:
Right now `manifest.webapp` is not included in the local deploy or the production build, so the app is not detected by the OWA module no matter how I try to deploy.

psbrandt commented on README.md:
`````diff
@@ -25,32 +25,33 @@ The following features are currently supported:
 ````
 
-Be sure to include the following in your `html` files at the position you want the Bower dependencies injected:
+Be sure to include the loading statement in your `.js` files at the modules you want your dependencies injected:
`````
Let's rewrite this sentence as:

> To use the installed package, import it as follows:

psbrandt commented on README.md:
```diff
@@ -25,32 +25,33 @@ The following features are currently supported:
+//import module
+require('package')
+//import and assign to variable
+import variableName from 'package';
```
I think we need to make a decision about whether we want to promote ES2015 (my preference) or whether we want to promote ES5. If we want to promote ES2015, then we should not show examples of the old syntax. We should also update the [template](https://github.com/psbrandt/generator-openmrs-owa/blob/master/app/templates/scripts/index.js) to look more like ES2015.

What do you think @rkorytkowski?

rkorytkowski replied:
I agree with promoting ES2015 (ES6) in the owa generator and using proper ES6 imports here. That said there are not yet many examples of code in ES6 so I expect ES5 and ES6 to interleave in our codebase. I do not think it's a bad thing and we should not force ourselves to always use the ES6 syntax. Similarly we do not force using the Java 8 syntax instead of Java 7. Let's allow for an evolution and not a revolution :)

psbrandt replied:
> we should not force ourselves to always use the ES6 syntax

Agreed. I just think we should have our examples in ES2015 to nudge people in the right direction 😄. That said, I am planning to [add ESLint to the generator](https://issues.openmrs.org/browse/OMRSJS-7) and using the Airbnb style (for non-Angular projects), which gives you warnings if you don't use ES2015.

PawelGutkowski replied:
> I am planning to add ESLint to the generator and using the Airbnb style (for non-Angular projects), which gives you warnings if you don't use ES2015.

In that case, I am removing old syntax example, to avoid confusement.

psbrandt commented on app/templates/README.md:
`````diff
@@ -41,7 +35,7 @@ which can be uploaded to the OpenMRS Open Web Apps module.
 
 ````
-gulp deploy-local
+npm build:deploy
`````
I guess you mean:

> npm **run** build:deploy

As mentioned in the comment at the bottom, both the prod build and the deploy do not work because the `manifest.webapp` file is not included.

psbrandt commented on app/templates/README.md:
```diff
@@ -70,25 +64,25 @@ will need the `APP_ENTRY_POINT` entry in your `config.json` file:
+//import module
+require('package')
+//import and assign to variable
+import variableName from 'package';
```
See my comments above about the these few lines ↑

psbrandt commented on app/templates/scripts/index.js:
```diff
@@ -9,6 +9,9 @@
  * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
  */
+require("../css/<%= appId %>.css");
+require("../index.html");
```
I don't think I understand what's going on here. Why are we requiring HTML and CSS in our JavaScript?

PawelGutkowski commented:
I've fixed packing `manifest.webapp`. I think if we would treat index.js mainly as entry point, importing all other webapp components there makes sense and is common practice (like eg. [here](https://github.com/preboot/angular-webpack/blob/master/src/app/app.js)). Anyway, I've moved HTML and CSS imports out of our index.js. Now .css file is separate entry point, but bundled together with .js. I'm waiting with fixing imports and template for definite decision.

rkorytkowski commented on app/templates/_package.json:
```diff
@@ -6,20 +6,39 @@
+    "build:deploy": "npm run clean && webpack --progress --colors --mode=dev --target=web",
+    "watch": "npm run clean && webpack --progress --colors --watch --mode=dev --target=web",
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "zip": "bestzip dist/<%= appId %>.zip dist/*"
```
There was an issue with putting zip in dist. Basically bestzip also tried to zip the zip. For that reason I put zip outside of dist. See https://github.com/rkorytkowski/openmrs-owa-conceptdictionary/commit/3b2de698270c87a65d406a2db03147fe1719d6a5

PawelGutkowski replied:
Thanks for heads up!

PawelGutkowski commented:
Hey @psbrandt, is there anything else You want me to amend?

PawelGutkowski commented:
In further testing I have found out that keeping 2 independent entry chunks (for css and js) in one bundle file is not the best idea, scripts didn't work. Creating separate bundle for css files fixed the problem.

psbrandt commented:
@PawelGutkowski I'll take another look at this today.

psbrandt commented on app/templates/_package.json:
```diff
@@ -6,20 +6,39 @@
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "clean": "rimraf dist && rimraf coverage*",
+    "build": "webpack --progress --colors --mode=production --target=web",
+    "build:prod": "npm run clean && npm run test && npm run build && npm run zip",
```
Can you add a `build:dev` script so that it's possible to build a non-minified version for debugging?

PawelGutkowski replied:
@psbrandt Done.

psbrandt commented:
The `/img/omrs-button.png` file mentioned in `manifest.webapp` is not included in the generated ZIP file, so when you install the app, the app icon is broken on the **Manage Apps** page.  I guess this is because Webpack is optimising by base64 encoding and inlining the image. I think (at least for the app icon) we should keep the original PNG.

PawelGutkowski commented:
I have overlooked that, fixed.

psbrandt commented:
When I install a package, it seems like Webpack is including it in `app.bundle.js`, but I expect it to be in `vendor.bundle.js`, since it's a third party library. So, if run:

```sh
npm install --save moment
```

and then, if my `app.js` looks like:

```js
import moment from 'moment';

(function () {
  'use strict';
  document.addEventListener("DOMContentLoaded", function(event) {
    console.log('OpenMRS Open Web App Started.');
    console.log(moment().format('YYYY-MM-DD'));
  });
}());

```

it works, but `moment` seems to be in the wrong file. Maybe my expectation is wrong?

psbrandt commented:
I'm still trying to figure out how to install CSS-only packages (e.g. [animate.css](https://daneden.github.io/animate.css/)), or packages that contain both CSS and JS (e.g. [Bootstrap](http://getbootstrap.com/)). Have you been able to get such examples working?

PawelGutkowski commented:
Did You declared moment in vendor entry point? the same as jquery/angular [is?](https://github.com/psbrandt/generator-openmrs-owa/pull/15/files#diff-3644bc993eb397faf5b67532f9ebe8daR119). It's [necessary](https://webpack.github.io/docs/list-of-plugins.html#2-explicit-vendor-chunk). About installing css, besides importing our openmrs-contrib-uicommons to openmrs-owa-conceptdictionary, I haven't try it yet. I will look into it.

psbrandt commented:
> Did You declared moment in vendor entry point?

I don't think so. All I'm doing is what I posted in the comment above. I guess you are referring to this section of `webpack.config.js`:

```js
entry: {
  app : `${__dirname}/app/js/owa.js`,
  css: `${__dirname}/app/css/owa.css`,
  vendor : [
    'angular'
  ]
},
```

Can you update the README to explain this requirement?

PawelGutkowski commented:
Yeah, that's the part I meant. I have updated README.

PawelGutkowski commented:
I go `animate.css` working by importing it at the end of our css file ` @import '../../node_modules/animate.css/animate.min.css'`. There's no js file in this package which would export stylesheets, so pretty importing like `import 'animate.css'` is not avalaible. I will try to install Bootstrap too.

PawelGutkowski commented:
Installing bootstrap by webpack is painful, but there's [bootstrap-webpack](https://www.npmjs.com/package/bootstrap-webpack), which makes it quite easy. From my little experience I have to say that unfortunately, a lots of libraries are not adjusted to be used with webpack, and need some tricks/wrap packages to be installed.

psbrandt commented:
With the latest change, the app icon is working if I do `npm run watch` or `npm run build:deploy`, but not if I actually do a production build and upload the app via the **Manage Apps** page. It seems that the ZIP file is generated as follows:

```
openmrs-owa-test
├── app.bundle.min.js
├── css.bundle.min.js
├── dist
│   └── img
│       └── omrs-button.png
├── index.html
├── manifest.webapp
└── vendor.bundle.js
```

Note the extra `dist` directory. Can you confirm this is also happening for you?

What is the reason that you are now [including `openmrs-owa-` in the `appId`](https://github.com/PawelGutkowski/generator-openmrs-owa/blob/master/app/index.js#L207) when generating the ZIP file?

PawelGutkowski commented:
Yes, I have the same zip structure. Thanks for noticing, we overlooked that issue in Concept Dictionary, too. I will fix it later. 
What's wrong with naming ZIP file with appId?

psbrandt commented:
> What's wrong with naming ZIP file with appId?

There is nothing wrong with using the `appId`, but you've changed the behaviour of the build to also include "`openmrs-owa-`", so I'm just wondering what the reason is. Also, OpenMRS modules are named like `htmlformentry-2.6.omod`, not `openmrs-module-htmlformentry.omod`, so I think we should stick to that convention. How about we remove the `openmrs-owa-` from the ZIP filename and include the version from `package.json` just like you guys have done with the [Concept Dictionary OWA](https://bintray.com/openmrs/owa/openmrs-owa-conceptdictionary/1.0.0-beta.1/view)?

I think after the ZIP file is corrected we can land this PR. Thanks for all the hard work! We've made it a bit  more complicated to add and use different packages (because Webpack makes it more complicated), so I think having good documentation and examples would be a good thing. I will try to find the time to work on that, but PRs will always be welcome :).

PawelGutkowski commented:
> but you've changed the behaviour of the build to also include "openmrs-owa-", so I'm just wondering what the reason  is 

On master branch `appId` in `package.json` already was including 'openmrs-owa', I didn't change it. Now I see that `gulpfile.js` had appId without this prefix, so that's where the difference come from.

I couldn't change `bestzip` behaviour, so now zipping is managed from `webpack.config` using `archiver`. I am sticking to naming convention, and include version in zip name. There's one catch - when You upload zip, the adress to app contains version, eg. `localhost:8080/openmrs/owa/stuffitystuff-v0.1.0/index.html`.

I'm happy to help. About documentation, we are planning to create video tutorial for development using this generator and openmrs-contrib-uicommons. If You got any certain ideas to improve documentation, ping me, I will certainly help!

psbrandt commented:
@rkorytkowski, @sunbiz, how do you feel about including the version in the URL? Doing it this way would allow us to have multiple versions of the same app installed.

rkorytkowski commented:
@psbrandt, personally I wouldn't include the version in the URL. It will be troublesome to handle when adding links to an app from other apps unless we can provide a default version with URL consistent between versions. I think the benefits do not justify the complexity behind supporting multiple versions of an app.

psbrandt commented:
Okay @rkorytkowski, that reasoning makes sense to me. When I download the [Concept Dictionary OWA](https://bintray.com/openmrs/owa/openmrs-owa-conceptdictionary/1.0.0-beta.1/view), the filename has a version in it (as required by Bintray), so I suspect that if I were to install it into the OWA Module, it will have that version in the URL.

How do you think we should handle this? Whatever strategy we choose, we should definitely bake it into the generator (if applicable) sooner rather than later.

I guess a few options are:

1. Ask people to rename the ZIP files themselves
2. Add functionality to the OWA App Store to exclude the version number from the filename when downloading
3. Add functionality to the OWA Module to exclude the version number either when serving or when extracting the ZIP

I think my preference is probably number 3 (when extracting).

rkorytkowski commented:
I think the filename should not matter at all to the OWA module. It should read app id from manifest.webapp and put it under correct url.

psbrandt commented:
Agreed, so in that case we don't need to change anything here, and I think we merge this if you agree.

rkorytkowski commented:
I agree with the merge.
